File: 2707. Extra Characters in a String/solution.cpp
class Solution {
private:

    int recursion(string &s,int idx,unordered_map<string,int>&mp,int *dp){

        if(idx == s.size())
            return 0;

        if(dp[idx]!=-1)
            return dp[idx];

        string curr = "";
        int currExtra = 0;  
        int minExtra = s.size();

        for(int i=idx;i<s.size();i++){

            curr += s[i];
            currExtra = mp.count(curr) ? 0 : curr.size();
            int nxtExtra = recursion(s,i+1,mp,dp);
            int totalExtra = currExtra + nxtExtra;
            minExtra = min(minExtra,totalExtra);
        }

        return dp[idx] = minExtra;
    }
    
public:
    int minExtraChar(string s, vector<string>& dictionary) {

        int dp[s.size()];
        memset(dp,-1,sizeof(dp));
        unordered_map<string,int>mp;

        for(auto &x : dictionary)
             mp[x]++;

        return recursion(s,0,mp,dp);
    }
};

File: 2707. Extra Characters in a String/solution.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
var minExtraChar = function(s, dictionary) {

    const recursion = (s, idx, mp, dp) => {
        if (idx === s.length) {
            return 0;
        }

        if (dp[idx] !== -1) {
            return dp[idx];
        }

        let curr = "";
        let currExtra = 0;
        let minExtra = s.length;

        for (let i = idx; i < s.length; i++) {
            curr += s[i];
            currExtra = mp.has(curr) ? 0 : curr.length;
            let nxtExtra = recursion(s, i + 1, mp, dp);
            let totalExtra = currExtra + nxtExtra;
            minExtra = Math.min(minExtra, totalExtra);
        }

        return dp[idx] = minExtra;
    };

    let dp = new Array(s.length).fill(-1);
    let mp = new Map();

    dictionary.forEach((word) => {
        mp.set(word, (mp.get(word) || 0) + 1);
    });

    return recursion(s, 0, mp, dp);
};
